### PR TITLE
🎨  one token endpoint

### PR DESCRIPTION
**Pull request by kirrg001**
refs #7562
- delete /authentication/ghost
- Ghost-Admin will use /authentication/token for all use cases (password, refresh token and ghost.org authorization code)
- add new grant_type `authorization_code`

We can't use the `oauthorize.exchange.authorizationCode` function, because this wraps the express req/res and passport needs access to it. Because Ghost is just forwarding the `authorization_code` to Ghost.org, it's OK to not use `oauthorize.exchange.authorizationCode`.

### CODE REVIEW

File: core/server/auth/authenticate.js [hunk 1]
```diff
@@ -105,29 +105,6 @@ authenticate = {
                 }));
             }
         )(req, res, next);
-    },
-
-    // ### Authenticate Ghost.org User
-    authenticateGhostUser: function authenticateGhostUser(req, res, next) {
```
Comment (kirrg001) [inline core/server/auth/authenticate.js]: this just moved to `core/server/auth/oauth.js`

File: core/server/auth/oauth.js [hunk 2]
```diff
@@ -85,6 +124,13 @@ oauth = {
         // access token on behalf of the user who authorized the code.
         oauthServer.exchange(oauth2orize.exchange.refreshToken({userProperty: 'client'},
             exchangeRefreshToken));
+
+        /**
+         * We forward the authorization code to Ghost.org, so no need to call exchange.authorizationCode
+         * exchange.authorizationCode wraps the req and res object, which is used by passport!
+         * See exchangeAuthorizationCode!
+         */
+        oauthServer.exchange('authorization_code', exchangeAuthorizationCode);
```
Comment (sebgie) [inline core/server/auth/oauth.js]: `exchange.authorizationCode` takes over client extraction which is missing in the code above.

File: core/server/auth/oauth.js [hunk 3]
```diff
@@ -62,6 +63,44 @@ function exchangePassword(client, username, password, scope, done) {
...
+
+        if (!user) {
+            return next(new errors.UnauthorizedError({
+                message: i18n.t('errors.middleware.auth.accessDenied')
+            }));
+        }
+
+        authenticationAPI.createTokens({}, {context: {client_id: req.client.id, user: user.id}})
+            .then(function then(response) {
+                spamPrevention.resetCounter(user.get('email'));
```
Comment (sebgie) [inline core/server/auth/oauth.js]: Authorization codes are exempt from spam prevention according to the change here: https://github.com/TryGhost/Ghost/pull/7571/commits/28d40946234082c029fd77496d07590e07ffed6c#diff-7b513b6d183ddea58102b60e32c8b251R30. So no need to reset the counter :).

**Review by sebgie — APPROVED** [review]
:+1: for moving the endpoint to /token. Makes a lot of sense!!

I would stick with using oauth2orizes structure of going through exchange.authorizationCode to do request verification. This isn't a change that is needed now! I might take a stab at that, if that's okay :).

Something to keep in mind: Once we allow 3rd party apps to have authenticated access to the API we need to distinguish between Ghost.org and password auth.

File: core/server/auth/oauth.js (hunk 3 above)
Comment (kirrg001) [inline core/server/auth/oauth.js]: cool thanks 👍 will update

Comment (kirrg001) [inline core/server/auth/oauth.js]: so true 💃

File: core/server/auth/oauth.js (hunk 2 above)
Comment (kirrg001) [inline core/server/auth/oauth.js]: will add client validation 👍 thanks

Comment (kirrg001) [inline core/server/auth/oauth.js]: hm actually it does already validate the client
https://github.com/TryGhost/Ghost/pull/7571/files#diff-e1be96888c62e807a8c07447f0e5f164R184

Comment (sebgie) [inline core/server/auth/oauth.js]: yes, it is ready to merge imo! As I said in the other comment, this is something for later :+1:

**Comment by kirrg001** [comment]
Thanks for review, I've added a very small commit to update my comment description!This description is hopefully a bit more clear for later.

https://github.com/TryGhost/Ghost/pull/7571/commits/4db8d4246c49a47c6847e61a23537d39ec3ef393

**Pull request merged**